### ISSUE TITLE
Rank histogram

### ISSUE THREAD

**Issue by SamAmco** [online-go/online-go.com#2431]
As a player it would be really nice to know what percentile I am in the global distribution of ranks on OGS. This has been raised before in the forum here : https://forums.online-go.com/t/rank-distribution-2021/33010

This could be as simple as a number e.g. "you are in the top 35% of players" or as comprehensive as an interactive histogram like the one found on lichess here : https://lichess.org/stat/rating/distribution/classical

**Comment by viczommers** [comment]
@anoek is there a better API socket than [v1 Django REST](https://apidocs.online-go.com/)? + doc/api_v1_doc.yaml

The best endpoint I found is: https://online-go.com/api/v1/players/?ordering=-rank&page=1&page_size=100 but I cant filter results by field without getting the whole result payload first.

I can write a simple scheduled job that will be traversing players list (once a day/week), fetching all rankings & plotting into frontend plot similar to `RatingsChart` inside `ProfileCard`. 

I just need either: 
(A.) an endpoint to store the json dump if we put it into the main client or 
(B.) I can just create a standalone FastAPI/Flask app, call it OGS stats...

**Comment by viczommers** [comment]
@anoek Sure, I digged around the Lichess repo and replicated their .js chart + more or less followed their histogram logic.
It fetches`ratings.jsonl`, transforms it into 2 arrays: freq, ratings with `createRatingDistribution()` and plots results. There is also a variable to display user's rating.

- Chart https://viczommers.github.io/Go-Rating-Distribution/  
- Code Repo https://github.com/viczommers/Go-Rating-Distribution/

The only major modification I made is that each frequency/count represents sum of all players with rating in range between the current Bin label & excludes previous Bin label (Upper Boundary). For example, the 1600 point shows all players rated between 1501-1600.

Lichess logic was other way around (Lower Boundary) - Bin labeled 1600 shows all players rated in [1600, 1699] range.

Below are some assumptions for MVP:
- Glicko2 Ratings below 100 & above 3000 are excluded from Hist (extreme values)
- Provisional, high Deviation ratings are excluded from Hist as well
- User's rating will be plotted even if it is provisional/ high Deviation or outside of [100,3000] range, but the rendering is clipped at axis min,max.
- Histogram only for 'Overall' category (inline with current `ratings-chart`) + possible Kuy/Dan label conversion toggle.

Ideas for Next Step:
I can refactor `createRatingDistribution()` into python for Celery. I am not really sure how the backend DB looks but if we grab only one key-value pair `JSON.parse(line).overall.rating` from each user - that should be >2M rows? + filter out provisional high RDs and maybe some other rules like exclude professionals, bots. That should be ok for Numpy to handle?

**Comment by anoek** [comment]
> that should be >2M rows? + filter out provisional high RDs and maybe some other rules like exclude professionals, bots. That should be ok for Numpy to handle? 

Yeah that should be fine. In that case I think what I need from you is a function that iterates over a list of `overall` objects, do whatever filtering you find appropriate, and construct yourself an array that you then do your numpy stuff with and ultimately create a dict/json blob that can later be retrieved. I'll take that and wire it up into celery and the db, and store the results somewhere where you can read them.

**Comment by viczommers** [comment]
👍 I wrote [hist-transform.py](https://github.com/viczommers/Go-Rating-Distribution/blob/079d40c98b4f075f041e83477c222ad00714a9b3/hist-transform.py) to just process list of `overall` and return bytes. I chose `orjson` its faster for read/write + outputs bytes. I can also add `asyncio/aiohttp` and full json parsing. Let me know!

**Comment by anoek** [comment]
Awesome, thanks for that. I've adapted it and put it in, it's up on beta and should update once per day.

https://beta.online-go.com/termination-api/rating-histogram

**Comment by viczommers** [comment]
sorry @anoek I cant see the json? is there a specific GET/POST?

**Comment by anoek** [comment]
Sorry another update got pushed up that didn't have that commit in there - it should be available now.

**Comment by viczommers** [comment]
@anoek Nice! 😎 this is just a static slice for now? (141 samples from 18-Match?).

The only other question left is UX where to place the hist:
- User container in ratings-row, below current ratings-chart or
- One of the menu Tabs (Community or Tools)
The [Frontend Hist](https://viczommers.github.io/Go-Rating-Distribution/) already has functionality for plotting player rating line so User container might be better, but at the same time it starts to feel a bit crammed and now all the users will have to scroll past it on their way to Active Games / Game History

**Comment by anoek** [comment]
>  this is just a static slice for now? (141 samples from 18-Match?).

It's the data from the beta server, much less than production.

Re where to place it, I'm not opinionated here - if you want to do a full chart that'd be welcome and probably appropriate under the tools drop down. To display what percentile the user is in, I don't have a great vision about where to put that, sounds like you have some good ideas to start from and see how they feel.

**Comment by SamAmco** [comment]
Thanks for the work on this guys, it's looking brilliant! I'm happy for you to close this issue if you're done with it.

**Comment by anoek** [comment]
Indeed, thanks again @viczommers !
